packages/gatsby/src/redux/actions/public.js

```diff
@@ -1300,4 +1302,35 @@ actions.createPageDependency = (
   }
 }
 
+/**
+ * Create a task. This is a long-running process that is generally
+ * started by a plugin.
+ *
+ * Gatsby doesn't finish its build process until all jobs are complete.
+ * @param {string} name The name of the job
+ * @param {Object} args An object of arguments the worker function will receive
+ * @example
+ * createTask({ name: `IMAGE_PROCESSING`, args: { inputPath: `something.jpeg`, transforms: [] } })
+ */
+
+actions.createTask = ({ name, args }, plugin) => dispatch => {
+  const id = uuid()
+  console.log(`createTask`)
+  const deferred = pDefer()
+  dispatch({
+    id,
+    type: `CREATE_TASK`,
+    name,
+    args,
+    plugin,
+    deferred,
+  })
+  return deferred.promise.finally(() => {
+    dispatch({
+      id,
+      type: `END_TASK`,
+    })
+  })
```

Comment:
For the future - let's make sure we don't rely on promises to dispatch `END_TASK`.

Currently in master/latest we can hit scenario in `gatsby develop` that some images won't be processed: 
If, after initial bootstrap, we update data, which will run queries that will queue image processing and we terminate process after query running and before image processing complete, Gatsby will not finish image processing on next run because it doesn't track jobs between runs (it's also impossible to re-run them with current jobs, as those are used only to tell gatsby there is long running work, but not how to run this)

We would need to persist tasks to our .cache so gatsby knows that it needs to complete them. To persist we need to make sure we only use serializable primitives (and promises are not serializable).

This also will enable us to finally implement image processing on demand in `gatsby develop` which will significantly improve DX (as we can queue jobs, and only ones that users need when navigating development site will be executed).

All of this doesn't need to happen in initial implementation, as this is not one of main goals of it, but let's make sure we are ready to do so in the future.

Reply:
Also when jobs are serializable — we can send jobs to the cloud to do

Reply:
Yup, they’re serializable now and are persisted between processes and re triggered if incomplete